// @loaders.gl/tables/dist/es6/index.js
export { default as Schema } from './lib/schema/schema';
export { default as Field } from './lib/schema/field';
export { deduceTableSchema } from './lib/schema/schema-utils';
export { default as TableBatchBuilder } from './lib/table/table-batch-builder';
export { default as RowTableBatch } from './lib/table/row-table-batch';
export { default as ColumnarTableBatch } from './lib/table/columnar-table-batch';
export { default as JSONLoader } from './json-loader';
export { default as XMLLoader } from './xml-loader';
export { default as AsyncQueue } from './lib/utils/async-queue';
export { getTypeInfo } from './lib/types/type-utils';
export { DataType, Null, Bool, Int, Int8, Int16, Int32, Int64, Uint8, Uint16, Uint32, Uint64, Float, Float16, Float32, Float64, Utf8, Date, DateDay, DateMillisecond, Time, TimeMillisecond, TimeSecond, Timestamp, TimestampSecond, TimestampMillisecond, TimestampMicrosecond, TimestampNanosecond, Interval, IntervalDayTime, IntervalYearMonth, FixedSizeList } from './lib/types/arrow-like/type';
