// @loaders.gl/tables/dist/es6/lib/table/columnar-table-batch.js
export default class ColumnarTableBatch {
  constructor(schema, options = {}) {
    this.schema = schema;
    this.batchSize = options.batchSize || 'auto';
    this.length = 0;
    this.allocated = 0;
    this.columns = null;
    this.isChunkComplete = false;
    this.reallocateColumns();
  }

  addRow(row) {
    this.reallocateColumns();

    for (const fieldName in row) {
      this.columns[fieldName][this.length] = row[fieldName];
    }

    this.length++;
  }

  chunkComplete() {
    this.isChunkComplete = true;
  }

  isFull() {
    if (this.batchSize === 'auto') {
      return this.isChunkComplete;
    }

    return this.length >= this.allocated;
  }

  getBatch(options = {}) {
    this.pruneColumns();
    const columns = Array.isArray(this.schema) ? this.columns : {};

    if (!Array.isArray(this.schema)) {
      for (const fieldName in this.schema) {
        const field = this.schema[fieldName];
        columns[field.name] = this.columns[field.index];
      }
    }

    this.columns = null;
    this.isChunkComplete = false;
    return {
      data: columns,
      schema: this.schema,
      length: this.length
    };
  }

  reallocateColumns() {
    if (this.length < this.allocated) {
      return;
    }

    this.allocated = this.allocated > 0 ? this.allocated *= 2 : this.batchSize;
    this.columns = [];

    for (const fieldName in this.schema) {
      const field = this.schema[fieldName];
      const ArrayType = field.type || Float32Array;
      const oldColumn = this.columns[field.index];

      if (oldColumn && ArrayBuffer.isView(oldColumn)) {
        const typedArray = new ArrayType(this.allocated);
        typedArray.set(oldColumn);
        this.columns[field.index] = typedArray;
      } else if (oldColumn) {
        oldColumn.length = this.allocated;
        this.columns[field.index] = oldColumn;
      } else {
        this.columns[field.index] = new ArrayType(this.allocated);
      }
    }
  }

  pruneColumns() {
    this.columns = this.columns.map(column => column.slice(0, this.length));
  }

}
